### DISCUSSION TITLE
Using ktxTexture_SetImageFromStdioStream and ktxTexture_SetImageFromMemory to create a ktx2 file from a png file.

### DISCUSSION THREAD

**Discussion by Galo32** [KhronosGroup/KTX-Software#964] [Q&A] [answered]
Hello, i'm trying to convert my png filles into ktx2 files using the mentioned functions, i followed the https://github.khronos.org/KTX-Software/libktx/index.html#writektx "Writing a KTX or KTX2 file" section wich only refers to ...SetImageFromMemory function, when i tried it, the src parametter wasn't working so i dig into the ktx.h file and found the ...SetImageFromStdioStrem wich already work, my questions are:
1- ktxTexture_SetImageFromMemory function expects the src parametter to point to memory with the **entire** file or only need the texels/pixels data? I mean, src= [PNG FILE] or src = [pixel0, pixel1, pixel2, ...].
2- What's about srcSize?
3- ktxTexture_SetImageFromStdioStream function expect the src parametter to be a valid FILE* or what?
4- What's about srcSize?

**Comment by MarkCallow** [comment, marked as answer]
> 1- ktxTexture_SetImageFromMemory function expects the src parametter to point to memory with the **entire** file or only need the texels/pixels data? I mean, src= [PNG FILE] or src = [pixel0, pixel1, pixel2, ...].

`src` is a pointer to the decoded image (i.e. the raw pixel data) in memory.

> 2- What's about srcSize?

It is the byte size of the image: width * height * bytes_per_pixel.

> 3- ktxTexture_SetImageFromStdioStream function expect the src parametter to be a valid FILE* or what?

A valid FILE* with its read pointer set to start of decoded image data. This function is not useful with a PNG file as you need to decode the data before passing it to libktx. There the data will already be loaded in memory.

> 4- What's about srcSize?

It is the same as for `ktxTexture_SetImageFromMemory`.

You can use `ktx create` from the tool suite to make a KTX file from PNG file(s).

**Reply by Galo32** [reply]
thanks you a lot! 
What do you mean with use ktx create tool? 
I want to make my program convert image files into ktx, this tool could help me?

**Reply by MarkCallow** [reply]
`ktx create` is a command line tool for creating KTX files from, among other inputs, PNG files. It won't help if you are writing a program to do that job, except as a source of sample code. It is one of several tools included in the release packages available from this project.
